api/cluster.go

```diff
@@ -106,6 +108,37 @@ func (s *Server) indexFind(ctx *middleware.Context, req models.IndexFind) {
 	response.Write(ctx, response.NewMsgp(200, resp))
 }
 
+func (s *Server) indexTagDetails(ctx *middleware.Context, req models.IndexTagDetails) {
+	values, err := s.MetricIndex.TagDetails(req.OrgId, req.Tag, req.Filter, req.From)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, err.Error()))
+		return
+	}
+	response.Write(ctx, response.NewMsgp(200, &models.IndexTagDetailsResp{Values: values}))
+}
+
+func (s *Server) indexTags(ctx *middleware.Context, req models.IndexTags) {
+	tags, err := s.MetricIndex.Tags(req.OrgId, req.Filter, req.From)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, err.Error()))
+		return
+	}
+	response.Write(ctx, response.NewMsgp(200, &models.IndexTagsResp{Tags: tags}))
+}
+
+func (s *Server) indexTagFindSeries(ctx *middleware.Context, req models.IndexTagFindSeries) {
+	metrics, err := s.MetricIndex.FindByTag(req.OrgId, req.Expressions, req.From)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, err.Error()))
+		return
+	}
+	metricNames := make([]string, 0, len(metrics))
+	for _, metric := range metrics {
+		metricNames = append(metricNames, metric)
+	}
```

Comment:
why the copy here?
it looks like this slice is generated from scratch in `(m *MemoryIdx) resolveIDs` so we can just reuse it safely.

Reply:
in general, I think we should also be a bit more precise around semantics of reuse/concurrent access. and whenever returning types that are/have pointers or references (pointers, slices, structs with slices), clarify what the caller is allowed or supposed to do.
for example for MetricIndex.FindByTag which returns a `[]string`. This concern shouldn't leak across api boundaries, i.e. we shouldn't have to go digging into index implementations to figure out if we need to copy or not.

I think we can also take inspiration from various packages (both in standard libraries and 3rd packages) and take this a step further. e.g.:
- https://golang.org/pkg/io/#Reader
- https://github.com/tinylib/msgp/blob/b2b6a672cf1e5b90748f79b8b81fc8c5cf0571a1/msgp/write.go#L90
We would have the caller pass in a `[]string`, which is then used to populate the response and then returned like in the msgp example. the benefit is then that the ownership of that slice is more obviously the caller's and the caller can also control allocations by using a pool. (the MetricIndex can't work with a pool directly because when it returns data it doesn't know when the data is no longer used. unless we add an extra callback function but that just complicates things)
if the caller doesn't care about any of this (and initially we often don't, until we profile and demonstrate we really need a pool or something) it can just pass `nil` as input arg.

so the implication here is that the first arg of `FindByTag` would be a `[]string` and we change the method documentation to say something like `	It stores the metric names that match the given conditions into the provided slice and returns the extended slice and any errors encountered.`

let me know if any of this doesn't make sense.

Reply:
Regarding passing the slice to store results in as an argument:
- I understand the benefit of that "implied" ownership/who's responsible for it

But I think apart from that there are more disadvantages than advantages:
- F.e. `resolvIDs` will preallocate the required number of elements in the result slice because it already knows them at the time the slice is initialized
- Even if the documentation says that the caller just needs to pass the slice to store results in, I think probably almost everybody who's going to use that will still look at the code of that called function anyway. 
- It just makes the function signature longer

Reply:
> F.e. resolvIDs will preallocate the required number of elements in the result slice because it already knows them at the time the slice is initialized

you can still do the equivalent with a slice that is passed in (whether it is nil or has any length already) see https://github.com/raintank/schema/blob/4a55004ab1f55549c476850030990da9f72574b5/metric.go#L72 however this is non-trivial, so i'm fine with leaving it like this for now (we can re-discuss when we need to introduce a pool or something)
but my point (I think) remains that these semantics of what the caller can expect around reuse of the slice, should be documented in the method docs. 
e.g. they can just say something like "the caller can do to the returned slice what it wants".
or should this be implied and only ever document cases where this is *not* the case?

Reply:
i think if somebody wants to know what the caller can do to the returned slice they'll simply check the code of the instantiating function. otherwise every single method in the whole code base would need to have comments that explain such things that are usually obvious when looking at the code.